lib/cocoapods-core/cdn_source.rb

```diff
@@ -0,0 +1,210 @@
+      # after the source was initialized, is considered fresh enough.
+      @startup_time = Time.new
+
+      @executor = Concurrent::ThreadPoolExecutor.new(
+        :min_threads => 50,
+        :max_threads => 200,
+        :max_queue => 0 # unbounded work queue
+      )
+      super(repo)
+    end
+
+    # @return [String] The URL of the source.
+    #
+    def url
+      @url ||= File.read(repo + '.url')
+    end
+
+    # @return [String] The type of the source.
+    #
+    def type
+      'CDN'
+    end
+
+    def refresh_metadata
+      if metadata.nil?
+        specs_dir.mkpath
+        download_file('CocoaPods-version.yml')
+
+        # Optimization: all the version index files in the local cache will in all probability be
+        # requested anyaway, better preload them in parallel.
+        loaders = repo.glob('*/**/*.txt').map { |f| f.relative_path_from(repo) }.map do |file|
+          Concurrent::Promise.execute(:executor => @executor) do
+            download_file(file)
+          end
+        end
+        Concurrent::Promise.zip(*loaders).wait!
+      end
+
+      super
+    end
+
+    # @return [Pathname] The directory where the specs are stored.
+    #
+    def specs_dir
+      @specs_dir ||= repo + 'Specs'
+    end
+
+    def download_file(partial_url)
+      file_remote_url = url + partial_url.to_s
+      path = repo + partial_url
+
+      if File.exist?(path) && @startup_time < File.mtime(path)
+        Pod::UI.message "CDN: #{name} Relative path: #{partial_url} modified during this run! Returning local"
+        return partial_url
+      end
+
+      path.dirname.mkpath
+
+      etag_path = path.sub_ext(path.extname + '.etag')
+
+      etag = File.read(etag_path) if File.exist?(etag_path)
+      Pod::UI.message "CDN: #{name} Relative path: #{partial_url}, has ETag? #{etag}" unless etag.nil?
+
+      response = REST.get(file_remote_url,
+                          'If-None-Match' => etag)
+      case response.status_code
+      when 304
+        Pod::UI.message "CDN: #{name} Relative path not modified: #{partial_url}"
+        # We need to update the file modification date, as it is later used for freshness
+        # optimization. See #initialize for more information.
+        FileUtils.touch path
+        partial_url
+      when 200
+        File.open(path, 'w') { |f| f.write(response.body) }
+
+        etag_new = response.headers['etag'].first if response.headers.include?('etag')
+        Pod::UI.message "CDN: #{name} Relative path downloaded: #{partial_url}, save ETag: #{etag_new}"
+        File.open(etag_path, 'w') { |f| f.write(etag_new) } unless etag_new.nil?
+        partial_url
+      end
+    end
+
+    def local_file(partial_url)
+      File.open("#{repo}/#{partial_url}") do |file|
+        yield file if block_given?
+      end
+    end
+
+    def local_file_exist?(partial_url)
+      File.exist?("#{repo}/#{partial_url}")
+    end
+
+    def pod_path_partial(pod_name)
```

Comment:
nit: `relative_pod_path`

Reply:
also can this be private?

Reply:
actually now I'm wondering if `relative_pod_path` should be added to the main `Source` class